File: EduAI-LMS-Frontend/src/services/AttendanceService.js
import api from "../api/axiosConfig";

const AttendanceService = {

    // =====================================================
    // GET ALL ATTENDANCE
    // =====================================================

    getAllAttendance() {
        return api.get("/attendance");
    },


    // =====================================================
    // GET ATTENDANCE BY ID
    // =====================================================

    getAttendance(id) {
        return api.get(`/attendance/${id}`);
    },


    // =====================================================
    // MARK ATTENDANCE
    // =====================================================

    markAttendance(studentId, courseId, attendance) {

        console.log("MARK ATTENDANCE REQUEST");
        console.log("Student ID:", studentId);
        console.log("Course ID:", courseId);
        console.log("Payload:", attendance);

        return api.post(
            `/attendance/student/${studentId}/course/${courseId}`,
            attendance
        );
    },


    // =====================================================
    // UPDATE ATTENDANCE
    // =====================================================

    updateAttendance(id, attendance) {

        console.log("UPDATE ATTENDANCE REQUEST");
        console.log("Attendance ID:", id);
        console.log("Payload:", attendance);

        return api.put(
            `/attendance/${id}`,
            attendance
        );
    },


    // =====================================================
    // DELETE ATTENDANCE
    // =====================================================

    deleteAttendance(id) {

        return api.delete(
            `/attendance/${id}`
        );
    },


    // =====================================================
    // GET STUDENT ATTENDANCE
    // =====================================================

    getStudentAttendance(studentId) {

        return api.get(
            `/attendance/student/${studentId}`
        );
    },


    // =====================================================
    // GET COURSE ATTENDANCE
    // =====================================================

    getCourseAttendance(courseId) {

        return api.get(
            `/attendance/course/${courseId}`
        );
    },


    // =====================================================
    // GET ATTENDANCE BY DATE
    // =====================================================

    getAttendanceByDate(date) {

        return api.get(
            `/attendance/date/${date}`
        );
    },


    // =====================================================
    // GET ATTENDANCE PERCENTAGE
    // =====================================================

    calculateAttendancePercentage(
        studentId,
        courseId
    ) {

        return api.get(
            `/attendance/percentage/${studentId}/${courseId}`
        );
    }

};

export default AttendanceService;

File: backend/src/main/java/backend/service/service/AttendanceService.java
package backend.service;




import backend.entity.Attendance;
import backend.entity.Course;
import backend.entity.User;
import backend.repository.AttendanceRepository;
import backend.repository.CourseRepository;
import backend.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
public class AttendanceService {

    @Autowired
    private AttendanceRepository attendanceRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private CourseRepository courseRepository;

    // ================= MARK ATTENDANCE =================

    public Attendance markAttendance(
            Long studentId,
            Long courseId,
            Attendance attendance) {

        User student = userRepository.findById(studentId)
                .orElseThrow(() -> new RuntimeException("Student not found"));

        Course course = courseRepository.findById(courseId)
                .orElseThrow(() -> new RuntimeException("Course not found"));

        if (attendanceRepository.existsByStudentIdAndCourseIdAndAttendanceDate(
                studentId,
                courseId,
                attendance.getAttendanceDate())) {

            throw new RuntimeException(
                    "Attendance already marked for this student on this date."
            );
        }

        attendance.setStudent(student);
        attendance.setCourse(course);

        return attendanceRepository.save(attendance);
    }

    // ================= UPDATE =================

    public Attendance updateAttendance(
            Long id,
            Attendance updatedAttendance) {

        Attendance attendance = attendanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Attendance not found"));

        attendance.setAttendanceDate(updatedAttendance.getAttendanceDate());
        attendance.setStatus(updatedAttendance.getStatus());
        attendance.setRemarks(updatedAttendance.getRemarks());

        return attendanceRepository.save(attendance);
    }

    // ================= DELETE =================

    public void deleteAttendance(Long id) {

        Attendance attendance = attendanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Attendance not found"));

        attendanceRepository.delete(attendance);
    }

    // ================= GET ALL =================

    public List<Attendance> getAllAttendance() {

        return attendanceRepository.findAll();
    }

    // ================= GET BY ID =================

    public Attendance getAttendanceById(Long id) {

        return attendanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Attendance not found"));
    }

    // ================= GET STUDENT =================

    public List<Attendance> getStudentAttendance(Long studentId) {

        return attendanceRepository.findByStudentId(studentId);
    }

    // ================= GET COURSE =================

    public List<Attendance> getCourseAttendance(Long courseId) {

        return attendanceRepository.findByCourseId(courseId);
    }

    // ================= GET DATE =================

    public List<Attendance> getAttendanceByDate(LocalDate date) {

        return attendanceRepository.findByAttendanceDate(date);
    }

    // ================= ATTENDANCE PERCENTAGE =================

    public double calculateAttendancePercentage(
            Long studentId,
            Long courseId) {

        List<Attendance> attendanceList =
                attendanceRepository.findByCourseId(courseId);

        long totalClasses = attendanceList.stream()
                .filter(a -> a.getStudent().getId().equals(studentId))
                .count();

        if (totalClasses == 0) {
            return 0;
        }

        long presentClasses = attendanceList.stream()
                .filter(a -> a.getStudent().getId().equals(studentId))
                .filter(a -> a.getStatus().name().equals("PRESENT"))
                .count();

        return (double) presentClasses / totalClasses * 100;
    }

}